File: src/database_transfer.py
from __future__ import annotations

import argparse
import base64
import hashlib
import json
import sqlite3
import stat
from collections.abc import Iterable, Mapping, Sequence
from pathlib import Path
from typing import Any

from .db import connect_database, initialize_database
from .db_protocol import DatabaseConnection
from .schema_scope import SCHEMA_SCOPE_INVENTORY


FORMAT_VERSION = 1
EXCLUDED_TABLE_PREFIXES = ("observation_fts", "sqlite_")


def _quote(identifier: str) -> str:
    return '"' + identifier.replace('"', '""') + '"'


def _encode(value: object) -> object:
    if isinstance(value, bytes):
        return {"$base64": base64.b64encode(value).decode("ascii")}
    return value


def _decode(value: object) -> object:
    if isinstance(value, Mapping) and set(value) == {"$base64"}:
        return base64.b64decode(str(value["$base64"]), validate=True)
    return value


def _canonical_row(values: Sequence[object]) -> str:
    return json.dumps(
        [_encode(value) for value in values],
        ensure_ascii=True,
        separators=(",", ":"),
    )


def _table_names(connection: sqlite3.Connection) -> list[str]:
    return [
        str(row[0])
        for row in connection.execute(
            """SELECT name FROM sqlite_master
               WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY name"""
        )
        if not str(row[0]).startswith(EXCLUDED_TABLE_PREFIXES)
    ]


def _table_metadata(
    connection: sqlite3.Connection, table: str,
) -> tuple[list[str], list[str], list[str]]:
    columns = connection.execute(f"PRAGMA table_info({_quote(table)})").fetchall()
    column_names = [str(row[1]) for row in columns]
    primary_key = [
        str(row[1])
        for row in sorted((row for row in columns if int(row[5]) > 0), key=lambda row: int(row[5]))
    ]
    dependencies = sorted({
        str(row[2])
        for row in connection.execute(f"PRAGMA foreign_key_list({_quote(table)})")
        if str(row[2]) != table
    })
    return column_names, primary_key, dependencies


def _ordered_rows(
    connection: Any, table: str, columns: Sequence[str], primary_key: Sequence[str],
) -> Iterable[Sequence[object]]:
    selected = ",".join(_quote(column) for column in columns)
    order_columns = primary_key or columns
    order = ",".join(_quote(column) for column in order_columns)
    return connection.execute(
        f"SELECT {selected} FROM {_quote(table)} ORDER BY {order}"
    )


def _checksum_rows(rows: Iterable[Sequence[object]]) -> tuple[str, int]:
    digest = hashlib.sha256()
    count = 0
    for row in rows:
        digest.update(_canonical_row(row).encode("utf-8"))
        digest.update(b"\n")
        count += 1
    return digest.hexdigest(), count


def _ownership_check(connection: Any, table: str) -> dict[str, object]:
    rule = SCHEMA_SCOPE_INVENTORY.get(table)
    if rule is None:
        return {"scope": "unclassified", "checked_rows": 0, "unowned_rows": 0}
    if rule.scope == "global" or rule.owner_column is None or rule.owner_table is None:
        return {"scope": rule.scope, "checked_rows": 0, "unowned_rows": 0}
    owner_column = _quote(rule.owner_column)
    owner_table = _quote(rule.owner_table)
    child_table = _quote(table)
    row = connection.execute(
        f"""SELECT COUNT(*),
                   SUM(CASE WHEN child.{owner_column} IS NOT NULL AND owner.id IS NULL THEN 1 ELSE 0 END)
            FROM {child_table} AS child
            LEFT JOIN {owner_table} AS owner ON owner.id=child.{owner_column}"""
    ).fetchone()
    return {
        "scope": rule.scope,
        "checked_rows": int(row[0]),
        "unowned_rows": int(row[1] or 0),
    }


def export_sqlite_database(
    source_path: Path,
    output_directory: Path,
    *,
    mark_source_read_only: bool = False,
) -> Path:
    source_path = source_path.resolve()
    output_directory.mkdir(parents=True, exist_ok=True)
    connection = sqlite3.connect(source_path)
    try:
        connection.execute("PRAGMA wal_checkpoint(TRUNCATE)")
        connection.execute("PRAGMA foreign_keys=ON")
        schema_row = connection.execute(
            "SELECT COALESCE(MAX(version),0) FROM schema_migrations"
        ).fetchone()
        foreign_key_rows = [list(row) for row in connection.execute("PRAGMA foreign_key_check")]
        tables: dict[str, dict[str, object]] = {}
        total_rows = 0
        for table in _table_names(connection):
            columns, primary_key, dependencies = _table_metadata(connection, table)
            data_path = output_directory / f"{table}.jsonl"
            digest = hashlib.sha256()
            row_count = 0
            with data_path.open("w", encoding="utf-8", newline="\n") as output:
                for row in _ordered_rows(connection, table, columns, primary_key):
                    line = _canonical_row(row)
                    output.write(line + "\n")
                    digest.update(line.encode("utf-8"))
                    digest.update(b"\n")
                    row_count += 1
            total_rows += row_count
            tables[table] = {
                "columns": columns,
                "primary_key": primary_key,
                "dependencies": dependencies,
                "row_count": row_count,
                "sha256": digest.hexdigest(),
                "ownership": _ownership_check(connection, table),
            }
        manifest = {
            "format_version": FORMAT_VERSION,
            "schema_version": int(schema_row[0]),
            "source": str(source_path),
            "tables": tables,
            "totals": {"tables": len(tables), "rows": total_rows},
            "orphan_check": {
                "count": len(foreign_key_rows),
                "rows": foreign_key_rows,
            },
        }
        manifest_path = output_directory / "manifest.json"
        manifest_path.write_text(
            json.dumps(manifest, indent=2, sort_keys=True) + "\n", encoding="utf-8"
        )
    finally:
        connection.close()
    if mark_source_read_only:
        source_path.chmod(source_path.stat().st_mode & ~(stat.S_IWUSR | stat.S_IWGRP | stat.S_IWOTH))
    return manifest_path


def _load_manifest(manifest_path: Path) -> dict[str, Any]:
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    if manifest.get("format_version") != FORMAT_VERSION:
        raise ValueError("unsupported database transfer manifest version")
    if not isinstance(manifest.get("tables"), dict):
        raise ValueError("database transfer manifest has no tables")
    if int(manifest.get("orphan_check", {}).get("count", -1)) != 0:
        raise ValueError("source database contains foreign-key orphans")
    unowned = [
        table
        for table, metadata in manifest["tables"].items()
        if int(metadata.get("ownership", {}).get("unowned_rows", 0)) != 0
    ]
    if unowned:
        raise ValueError(f"source database contains unowned rows: {', '.join(sorted(unowned))}")
    unclassified = [
        table
        for table, metadata in manifest["tables"].items()
        if metadata.get("ownership", {}).get("scope") == "unclassified"
    ]
    if unclassified:
        raise ValueError(
            f"source database contains unclassified tables: {', '.join(sorted(unclassified))}"
        )
    return manifest


def _foreign_key_order(tables: Mapping[str, Mapping[str, object]]) -> list[str]:
    remaining = set(tables)
    ordered: list[str] = []
    while remaining:
        ready = sorted(
            table
            for table in remaining
            if not (set(tables[table].get("dependencies", ())) & remaining)
        )
        if not ready:
            raise ValueError(f"cyclic transfer dependencies: {', '.join(sorted(remaining))}")
        ordered.extend(ready)
        remaining.difference_update(ready)
    return ordered


def _verify_target(
    connection: DatabaseConnection, manifest: Mapping[str, Any],
) -> dict[str, object]:
    mismatches: list[str] = []
    total_rows = 0
    for table, metadata in manifest["tables"].items():
        columns = list(metadata["columns"])
        primary_key = list(metadata["primary_key"])
        checksum, row_count = _checksum_rows(
            _ordered_rows(connection, table, columns, primary_key)
        )
        total_rows += row_count
        if row_count != int(metadata["row_count"]):
            mismatches.append(f"{table}: row count")
        if checksum != metadata["sha256"]:
            mismatches.append(f"{table}: checksum")
        ownership = _ownership_check(connection, table)
        if int(ownership["unowned_rows"]) != 0:
            mismatches.append(f"{table}: tenant ownership")
    invalid_constraints = int(connection.execute(
        "SELECT COUNT(*) FROM pg_constraint WHERE NOT convalidated"
    ).fetchone()[0])
    if invalid_constraints:
        mismatches.append("unvalidated PostgreSQL constraints")
    if total_rows != int(manifest["totals"]["rows"]):
        mismatches.append("source/target row totals")
    if mismatches:
        raise ValueError("database transfer verification failed: " + ", ".join(mismatches))
    return {
        "schema_version": int(manifest["schema_version"]),
        "tables": len(manifest["tables"]),
        "rows": total_rows,
        "constraints_validated": invalid_constraints == 0,
    }


def import_postgresql_database(
    manifest_path: Path,
    database_url: str,
    *,
    replace_target: bool = False,
) -> dict[str, object]:
    if not replace_target:
        raise ValueError("PostgreSQL import requires explicit replace_target=True")
    manifest = _load_manifest(manifest_path)
    tables: dict[str, dict[str, object]] = manifest["tables"]
    order = _foreign_key_order(tables)
    connection = connect_database(database_url)
    try:
        initialize_database(connection)
        quoted_tables = ",".join(_quote(table) for table in order)
        connection.execute(f"TRUNCATE TABLE {quoted_tables} RESTART IDENTITY CASCADE")
        for table in order:
            metadata = tables[table]
            columns = list(metadata["columns"])
            placeholders = ",".join("?" for _ in columns)
            statement = (
                f"INSERT INTO {_quote(table)} "
                f"({','.join(_quote(column) for column in columns)}) VALUES ({placeholders})"
            )
            rows = []
            data_path = manifest_path.parent / f"{table}.jsonl"
            with data_path.open(encoding="utf-8") as source:
                for line in source:
                    rows.append(tuple(_decode(value) for value in json.loads(line)))
            if rows:
                connection.executemany(statement, rows)
        identity_rows = connection.execute(
            """SELECT table_name,column_name FROM information_schema.columns
               WHERE table_schema='public' AND is_identity='YES' ORDER BY table_name"""
        ).fetchall()
        for table, column in identity_rows:
            connection.execute(
                f"""SELECT setval(
                        pg_get_serial_sequence('{table}','{column}'),
                        COALESCE((SELECT MAX({_quote(str(column))}) FROM {_quote(str(table))}),1),
                        EXISTS(SELECT 1 FROM {_quote(str(table))})
                    )"""
            )
        result = _verify_target(connection, manifest)
        connection.commit()
        return result
    except Exception:
        connection.rollback()
        raise
    finally:
        connection.close()


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Transfer QBot4K SQLite data to PostgreSQL")
    subparsers = parser.add_subparsers(dest="command", required=True)
    export_parser = subparsers.add_parser("export")
    export_parser.add_argument("source", type=Path)
    export_parser.add_argument("output", type=Path)
    export_parser.add_argument("--mark-source-read-only", action="store_true")
    import_parser = subparsers.add_parser("import")
    import_parser.add_argument("manifest", type=Path)
    import_parser.add_argument("database_url")
    import_parser.add_argument("--replace-target", action="store_true")
    arguments = parser.parse_args(argv)
    if arguments.command == "export":
        path = export_sqlite_database(
            arguments.source,
            arguments.output,
            mark_source_read_only=arguments.mark_source_read_only,
        )
        print(path)
    else:
        result = import_postgresql_database(
            arguments.manifest,
            arguments.database_url,
            replace_target=arguments.replace_target,
        )
        print(json.dumps(result, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/ops/database_transfer.ts
import { Buffer } from "node:buffer";
import { createHash } from "node:crypto";
import { DatabaseSync } from "node:sqlite";
import postgres from "postgres";
import type { DatabaseParameter } from "../data/database.ts";
import { SCHEMA_SCOPE_INVENTORY } from "../data/schema_scope.ts";

export const DATABASE_TRANSFER_FORMAT_VERSION = 1;
const EXCLUDED_TABLE_PREFIXES = ["observation_fts", "sqlite_"] as const;
type TransferValue = null | string | number | bigint | Uint8Array;

export interface TransferOwnership {
  readonly scope: string;
  readonly checked_rows: number;
  readonly unowned_rows: number;
}

export interface TransferTable {
  readonly columns: readonly string[];
  readonly primary_key: readonly string[];
  readonly dependencies: readonly string[];
  readonly row_count: number;
  readonly sha256: string;
  readonly ownership: TransferOwnership;
}

export interface TransferManifest {
  readonly format_version: number;
  readonly schema_version: number;
  readonly source: string;
  readonly tables: Readonly<Record<string, TransferTable>>;
  readonly totals: { readonly tables: number; readonly rows: number };
  readonly orphan_check: {
    readonly count: number;
    readonly rows: readonly (readonly unknown[])[];
  };
}

export interface TransferResult {
  readonly schema_version: number;
  readonly tables: number;
  readonly rows: number;
  readonly constraints_validated: boolean;
}

export async function exportSqliteDatabase(
  sourcePath: string,
  outputDirectory: string,
  markSourceReadOnly = false,
): Promise<string> {
  const source = await Deno.realPath(sourcePath);
  await Deno.mkdir(outputDirectory, { recursive: true });
  const database = new DatabaseSync(source);
  const tables: Record<string, TransferTable> = {};
  let totalRows = 0;
  try {
    database.exec("PRAGMA wal_checkpoint(TRUNCATE)");
    database.exec("PRAGMA foreign_keys=ON");
    const schema = database.prepare(
      "SELECT COALESCE(MAX(version),0) AS version FROM schema_migrations",
    ).get() as Record<string, unknown>;
    const orphanRows = rows(database, "PRAGMA foreign_key_check").map((row) =>
      Object.values(row)
    );
    const tableNames = rows(
      database,
      "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY name",
    ).map((row) => String(row.name)).filter((table) =>
      !EXCLUDED_TABLE_PREFIXES.some((prefix) => table.startsWith(prefix))
    );
    for (const table of tableNames) {
      const metadata = tableMetadata(database, table);
      const dataPath = `${outputDirectory}/${table}.jsonl`;
      const output = await Deno.open(dataPath, {
        create: true,
        write: true,
        truncate: true,
      });
      const digest = createHash("sha256");
      let rowCount = 0;
      try {
        const selected = metadata.columns.map(quote).join(",");
        const ordered =
          (metadata.primaryKey.length ? metadata.primaryKey : metadata.columns)
            .map(quote).join(",");
        for (
          const row of rows(
            database,
            `SELECT ${selected} FROM ${quote(table)} ORDER BY ${ordered}`,
          )
        ) {
          const line = canonicalRow(
            metadata.columns.map((column) => row[column] as TransferValue),
          ) + "\n";
          const bytes = new TextEncoder().encode(line);
          await output.write(bytes);
          digest.update(bytes);
          rowCount += 1;
        }
      } finally {
        output.close();
      }
      totalRows += rowCount;
      tables[table] = Object.freeze({
        columns: Object.freeze(metadata.columns),
        primary_key: Object.freeze(metadata.primaryKey),
        dependencies: Object.freeze(metadata.dependencies),
        row_count: rowCount,
        sha256: digest.digest("hex"),
        ownership: ownershipCheck(database, table),
      });
    }
    const manifest: TransferManifest = Object.freeze({
      format_version: DATABASE_TRANSFER_FORMAT_VERSION,
      schema_version: Number(schema.version),
      source,
      tables: Object.freeze(tables),
      totals: Object.freeze({ tables: tableNames.length, rows: totalRows }),
      orphan_check: Object.freeze({
        count: orphanRows.length,
        rows: Object.freeze(orphanRows),
      }),
    });
    const manifestPath = `${outputDirectory}/manifest.json`;
    await Deno.writeTextFile(
      manifestPath,
      `${JSON.stringify(manifest, null, 2)}\n`,
    );
    if (markSourceReadOnly) {
      const mode = (await Deno.stat(source)).mode;
      if (mode !== null) await Deno.chmod(source, mode & ~0o222);
    }
    return manifestPath;
  } finally {
    database.close();
  }
}

export async function importPostgresDatabase(
  manifestPath: string,
  databaseUrl: string,
  replaceTarget = false,
): Promise<TransferResult> {
  if (!replaceTarget) {
    throw new TypeError(
      "PostgreSQL import requires explicit replace_target=true",
    );
  }
  const manifest = await loadManifest(manifestPath);
  const order = foreignKeyOrder(manifest.tables);
  const sql = postgres(databaseUrl, { max: 1, onnotice: () => undefined });
  try {
    const schemaSql = await Deno.readTextFile(
      new URL("../postgres_schema.sql", import.meta.url),
    );
    await sql.unsafe(schemaSql);
    return await sql.begin(async (transaction) => {
      if (order.length) {
        await transaction.unsafe(
          `TRUNCATE TABLE ${
            order.map(quote).join(",")
          } RESTART IDENTITY CASCADE`,
        );
      }
      for (const table of order) {
        const metadata = manifest.tables[table];
        const columnSql = metadata.columns.map(quote).join(",");
        for await (
          const batch of jsonLineBatches(
            new URL(`./${table}.jsonl`, toDirectoryUrl(manifestPath)),
          )
        ) {
          if (!batch.length) continue;
          const parameters = batch.flatMap((values) => values.map(decodeValue));
          const width = metadata.columns.length;
          const valueSql = batch.map((_, rowIndex) =>
            `(${
              metadata.columns.map((_, columnIndex) =>
                `$${rowIndex * width + columnIndex + 1}`
              ).join(",")
            })`
          ).join(",");
          await transaction.unsafe(
            `INSERT INTO ${quote(table)} (${columnSql}) VALUES ${valueSql}`,
            parameters,
          );
        }
      }
      const identities = await transaction.unsafe(
        `SELECT table_name,column_name FROM information_schema.columns
          WHERE table_schema='public' AND is_identity='YES' ORDER BY table_name`,
      );
      for (const identity of identities) {
        const table = String(identity.table_name);
        const column = String(identity.column_name);
        await transaction.unsafe(
          `SELECT setval(pg_get_serial_sequence($1,$2),
             COALESCE((SELECT MAX(${quote(column)}) FROM ${quote(table)}),1),
             EXISTS(SELECT 1 FROM ${quote(table)}))`,
          [table, column],
        );
      }
      return await verifyTarget(transaction, manifest);
    });
  } finally {
    await sql.end();
  }
}

export function canonicalRow(values: readonly TransferValue[]): string {
  const json = JSON.stringify(values.map(encodeValue));
  let ascii = "";
  for (let index = 0; index < json.length; index += 1) {
    const code = json.charCodeAt(index);
    ascii += code > 0x7f
      ? `\\u${code.toString(16).padStart(4, "0")}`
      : json[index];
  }
  return ascii;
}

function encodeValue(value: TransferValue): unknown {
  if (value instanceof Uint8Array) {
    return { $base64: Buffer.from(value).toString("base64") };
  }
  if (typeof value === "bigint") return Number(value);
  return value;
}

function decodeValue(value: unknown): DatabaseParameter {
  if (
    value && typeof value === "object" && !Array.isArray(value) &&
    Object.keys(value).length === 1 && "$base64" in value
  ) {
    return Buffer.from(
      String((value as { $base64: unknown }).$base64),
      "base64",
    );
  }
  if (
    value === null || typeof value === "string" ||
    typeof value === "number" || typeof value === "boolean"
  ) {
    return value;
  }
  throw new TypeError("database transfer row contains an unsupported value");
}

function rows(database: DatabaseSync, sql: string): Record<string, unknown>[] {
  return database.prepare(sql).all() as Record<string, unknown>[];
}

function quote(identifier: string): string {
  return `"${identifier.replaceAll('"', '""')}"`;
}

function tableMetadata(database: DatabaseSync, table: string): {
  columns: string[];
  primaryKey: string[];
  dependencies: string[];
} {
  const columns = rows(database, `PRAGMA table_info(${quote(table)})`);
  return {
    columns: columns.map((column) => String(column.name)),
    primaryKey: columns.filter((column) => Number(column.pk) > 0)
      .sort((left, right) => Number(left.pk) - Number(right.pk))
      .map((column) => String(column.name)),
    dependencies: [
      ...new Set(
        rows(database, `PRAGMA foreign_key_list(${quote(table)})`)
          .map((foreignKey) => String(foreignKey.table))
          .filter((dependency) => dependency !== table),
      ),
    ].sort(),
  };
}

function ownershipCheck(
  database: DatabaseSync,
  table: string,
): TransferOwnership {
  const rule = SCHEMA_SCOPE_INVENTORY[table];
  if (!rule) {
    return { scope: "unclassified", checked_rows: 0, unowned_rows: 0 };
  }
  if (!rule.ownerColumn || !rule.ownerTable) {
    return { scope: rule.scope, checked_rows: 0, unowned_rows: 0 };
  }
  const result = database.prepare(
    `SELECT COUNT(*) AS checked_rows,
       SUM(CASE WHEN child.${quote(rule.ownerColumn)} IS NOT NULL
                    AND owner.id IS NULL THEN 1 ELSE 0 END) AS unowned_rows
       FROM ${quote(table)} AS child
       LEFT JOIN ${quote(rule.ownerTable)} AS owner
         ON owner.id=child.${quote(rule.ownerColumn)}`,
  ).get() as Record<string, unknown>;
  return {
    scope: rule.scope,
    checked_rows: Number(result.checked_rows),
    unowned_rows: Number(result.unowned_rows ?? 0),
  };
}

async function loadManifest(path: string): Promise<TransferManifest> {
  const manifest = JSON.parse(
    await Deno.readTextFile(path),
  ) as TransferManifest;
  if (manifest.format_version !== DATABASE_TRANSFER_FORMAT_VERSION) {
    throw new TypeError("unsupported database transfer manifest version");
  }
  if (!manifest.tables || typeof manifest.tables !== "object") {
    throw new TypeError("database transfer manifest has no tables");
  }
  if (Number(manifest.orphan_check?.count ?? -1) !== 0) {
    throw new TypeError("source database contains foreign-key orphans");
  }
  const unowned = Object.entries(manifest.tables)
    .filter(([, metadata]) => metadata.ownership.unowned_rows !== 0)
    .map(([table]) => table).sort();
  if (unowned.length) {
    throw new TypeError(
      `source database contains unowned rows: ${unowned.join(", ")}`,
    );
  }
  const unclassified = Object.entries(manifest.tables)
    .filter(([, metadata]) => metadata.ownership.scope === "unclassified")
    .map(([table]) => table).sort();
  if (unclassified.length) {
    throw new TypeError(
      `source database contains unclassified tables: ${
        unclassified.join(", ")
      }`,
    );
  }
  return manifest;
}

export function foreignKeyOrder(
  tables: Readonly<Record<string, TransferTable>>,
): readonly string[] {
  const remaining = new Set(Object.keys(tables));
  const ordered: string[] = [];
  while (remaining.size) {
    const ready = [...remaining].filter((table) =>
      !tables[table].dependencies.some((dependency) =>
        remaining.has(dependency)
      )
    ).sort();
    if (!ready.length) {
      throw new TypeError(
        `cyclic transfer dependencies: ${[...remaining].sort().join(", ")}`,
      );
    }
    ordered.push(...ready);
    ready.forEach((table) => remaining.delete(table));
  }
  return Object.freeze(ordered);
}

async function verifyTarget(
  connection: postgres.TransactionSql,
  manifest: TransferManifest,
): Promise<TransferResult> {
  const mismatches: string[] = [];
  let totalRows = 0;
  for (const [table, metadata] of Object.entries(manifest.tables)) {
    const columnTypes = await connection.unsafe(
      `SELECT column_name,data_type FROM information_schema.columns
        WHERE table_schema='public' AND table_name=$1`,
      [table],
    );
    const bigintColumns = new Set(
      columnTypes.filter((column) => column.data_type === "bigint")
        .map((column) => String(column.column_name)),
    );
    const selected = metadata.columns.map(quote).join(",");
    const ordered =
      (metadata.primary_key.length ? metadata.primary_key : metadata.columns)
        .map(quote).join(",");
    const targetRows = await connection.unsafe(
      `SELECT ${selected} FROM ${quote(table)} ORDER BY ${ordered}`,
    );
    const digest = createHash("sha256");
    for (const row of targetRows) {
      digest.update(
        `${
          canonicalRow(
            metadata.columns.map((column) => {
              const value = row[column] as TransferValue;
              return bigintColumns.has(column) && typeof value === "string"
                ? Number(value)
                : value;
            }),
          )
        }\n`,
      );
    }
    totalRows += targetRows.length;
    if (targetRows.length !== metadata.row_count) {
      mismatches.push(`${table}: row count`);
    }
    if (digest.digest("hex") !== metadata.sha256) {
      mismatches.push(`${table}: checksum`);
    }
    const rule = SCHEMA_SCOPE_INVENTORY[table];
    if (rule?.ownerColumn && rule.ownerTable) {
      const [ownership] = await connection.unsafe(
        `SELECT COUNT(*) AS unowned_rows FROM ${quote(table)} AS child
          LEFT JOIN ${quote(rule.ownerTable)} AS owner
            ON owner.id=child.${quote(rule.ownerColumn)}
         WHERE child.${
          quote(rule.ownerColumn)
        } IS NOT NULL AND owner.id IS NULL`,
      );
      if (Number(ownership.unowned_rows) !== 0) {
        mismatches.push(`${table}: tenant ownership`);
      }
    }
  }
  const [constraints] = await connection.unsafe(
    "SELECT COUNT(*) AS count FROM pg_constraint WHERE NOT convalidated",
  );
  if (Number(constraints.count) !== 0) {
    mismatches.push("unvalidated PostgreSQL constraints");
  }
  if (totalRows !== manifest.totals.rows) {
    mismatches.push("source/target row totals");
  }
  if (mismatches.length) {
    throw new TypeError(
      `database transfer verification failed: ${mismatches.join(", ")}`,
    );
  }
  return Object.freeze({
    schema_version: manifest.schema_version,
    tables: Object.keys(manifest.tables).length,
    rows: totalRows,
    constraints_validated: true,
  });
}

async function* jsonLineBatches(
  url: URL,
  batchSize = 500,
): AsyncGenerator<unknown[][]> {
  const file = await Deno.open(url, { read: true });
  let pending = "";
  let batch: unknown[][] = [];
  for await (
    const chunk of file.readable.pipeThrough(new TextDecoderStream())
  ) {
    pending += chunk;
    const lines = pending.split("\n");
    pending = lines.pop() ?? "";
    for (const line of lines) {
      if (!line) continue;
      batch.push(JSON.parse(line) as unknown[]);
      if (batch.length === batchSize) {
        yield batch;
        batch = [];
      }
    }
  }
  if (pending.trim()) batch.push(JSON.parse(pending) as unknown[]);
  if (batch.length) yield batch;
}

function toDirectoryUrl(path: string): URL {
  const url = new URL(path, `file://${Deno.cwd()}/`);
  return new URL("./", url);
}
